Publish odometry and current steering angle.

TorBorve opened the pull request:
Hi!
I'm using your repo to simulate an [autonomous car](https://www.fuelfighter.no/) . My [control system](https://github.com/TorBorve/mpc_local_planner) needs to know the current steering angle of the car to work optimally. Therefore I added a publisher that publishes the current steering angle on the wheels. Also I added a odom publisher since often SLAM algorithms require a odom topic. Lastly I added a front axle link. This was originally because it can be nice to look at the position of the front axle instead of rear axle. However, iẗ́'s not being used at the moment so it could also be removed.

Feel free to come with feedback!

robustify commented on audibot_description/urdf/audibot.urdf.xacro:
```diff
@@ -146,6 +146,19 @@
+    <parent link="base_footprint"/>
+    <child link="front_axle"/>
+  </joint>
+
```
I don't agree with adding this link and joint to the base URDF model because it does not serve any purpose for the behavior of the simulated vehicle.

Instead, I would recommend this:

1. Create a new URDF model in a separate ROS package
2. Include this base URDF model in the new one
4. Add custom links, sensors, etc. specific to your simulation goals
5. Write a modified `audibot_robot.launch` to load your new custom URDF model into Gazebo

Such a custom URDF file would look something like this:
```
<robot name="custom_vehicle_sim" xmlns:xacro="http://www.ros.org/wiki/xacro" >
  <xacro:include filename="$(find audibot_description)/urdf/audibot.urdf.xacro" />

  <link name="front_axle">
    <inertial>
      <mass value="0.1"/>
      <inertia ixx="0.0001" ixy="0.0001" ixz="0.0001" iyy="0.0001" iyz="0.0001" izz="0.0001" />
    </inertial>
  </link>

  <joint name="front_axle_joint" type="fixed">
    <origin xyz="2.652 0 0" rpy="0 0 0"/>
    <parent link="base_footprint"/>
    <child link="front_axle"/>
  </joint>

  <gazebo reference="front_axle_joint" >
    <preserveFixedJoint>true</preserveFixedJoint>
  </gazebo>
</robot>
```

robustify commented on audibot_description/urdf/audibot.urdf.xacro:
```diff
@@ -178,4 +191,8 @@
+  <!-- prevent fixed joint reduction -->
+  <gazebo reference="front_axle_joint" >
+    <preserveFixedJoint>true</preserveFixedJoint>
+  </gazebo>
```
Move this to custom URDF model too

robustify commented on audibot_gazebo/src/AudibotInterfacePlugin.cpp:
```diff
@@ -250,6 +257,24 @@ void AudibotInterfacePlugin::feedbackTimerCallback(const ros::TimerEvent& event)
+  pub_odom_.publish(odom_msg);
+
+  std_msgs::Float64 steering;
+  steering.data = current_steering_angle_;
```
I agree that the simulator should output the current steering angle because many control strategies can use it and because any car that is controllable by a computer would have that information available. However, production vehicles usually measure the steering wheel angle, not the kinematic steering angle of the tires with the road.

Instead of publishing the kinematic steering angle, please multiply this by `AUDIBOT_STEERING_RATIO` to publish the steering wheel angle instead.

robustify reviewed (changes requested):
Hi @TorBorve,

Thanks for your contribution! I'm glad you're finding the simulator helpful; it looks like you are doing some cool things!

See my other review comments on your changes for some feedback. I like the odometry and steering angle outputs, but not the front axle link. Let me know what you think.

TorBorve commented:
Thanks for your constructive feedback!

I agree that we can remove the front_axle link. As you said it does not serve any purpose for the simulated car. It would be better to make your own urdf as you said. Therefore, I removed the front_axle.

It is a bit annoying to multiply with `AUDIBOT_STEERING_RATIO` and then dividing with it later in the control system. However, it should be the same as `/steering_cmd` and if that is the standard then I’m fine with that. I also changed the name to `/steering_state` as that is the convention used with `/gear_state`.

I also added information about the new topics in the readme.

robustify reviewed (approved):
Looks good to me! Thanks for updating the README as well!

TorBorve commented:
No problem! I'm happy to help :)
